### PR TITLE
NET: Yahya/5143-topology-readme

### PR DESCRIPTION
**Pull request by yhassanzadeh13**
This PR drafts a readme for `/network/topology` package that defines its terminologies and overviews our topology construction approaches. It also adds some minor refactors to topology implementations by adding some loggings.

### CODE REVIEW

File: network/topology/randomizedTopology.go [hunk 1]
```diff
@@ -65,6 +72,7 @@ func (r RandomizedTopology) GenerateFanout(ids flow.IdentityList) (flow.Identity
 		// no subscribed channel id, hence skip topology creation
 		// we do not return an error at this state as invocation of MakeTopology may happen before
 		// node subscribing to all its channels.
+		r.logger.Warn().Msg("skips generating fanout with no subscribed channels")
```
Comment (yhassanzadeh13) [inline network/topology/randomizedTopology.go:75]: except for node startup, this is an indicator of topology disconnection on the network.

File: network/topology/README.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
+# Topology
+
+## Definitions
+
+In Flow, the term _topology_ captures the distributed protocol by which a node determines its fanout set. The _fanout set_ of a node is a subset of
+other nodes in the system by which the node interacts in the event of epidemic dissemination of information. The epidemic dissemination happens when a
```
Comment (vishalchangrani) [inline network/topology/README.md:6]:
```suggestion
other nodes in the system with whom the node interacts in the event of epidemic dissemination of information. The epidemic dissemination happens when a
```

File: network/topology/README.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+node _multicasts_ or _broadcasts_ a message. The former refers to the event when a node sends a message targeted for a subgroup of nodes in the
+system, while the latter refers to the event when a node aims at sending a message to the entire system. Note that the communications over the fanout
```
Comment (vishalchangrani) [inline network/topology/README.md:8]:
```suggestion
system, while the latter refers to the event when a node aims at sending a message to the entire system. (Topology algorithms are not used for direct one-to-one communication). Note that the communications over the fanout
```

File: network/topology/README.md [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+construction protocol should result in a connected graph with a very high probability over the life-time of Flow, which theoretically is
+infinitely-many epochs. In practice, we consider it as 2^30 many epochs. 
+
+The figure below shows an example of a topology graph with 7 nodes in the Flow network, which is a directed and connected graph. There is a path 
+between
+every two nodes and traversing the topology graph with BFS or DFS visits all the vertices. Also, the fanout set of each node is colored the same as
+the node itself, e.g., the fanout set of the red node is illustrated using red edges from it to other nodes of the network. Also, in this example,
+every node has a fanout size of 3.
+
+<p align="center">
```
Comment (vishalchangrani) [inline network/topology/README.md:23]: 👍  thanks for creating these images!

File: network/topology/README.md [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  <img src="topicBasedTopology.svg" alt="drawing" width="600"/>
+</p>
+
+### [RandomizedTopology](../../network/topology/randomizedTopology.go)
+
+The randomized-topology is similar to the topic-based topology, except that upon constructing a graph component for a topic, instead of choosing a 
+fixed-size fanout per node, it selects the fanout of the node probabilistically using an _edge probability_, `p`. 
+For the set of nodes subscribed to a topic, each node has any other node in its fanout with the probability of `p`, hence, the expected fanout of
```
Comment (vishalchangrani) [inline network/topology/README.md:52]:
```suggestion
For the set of nodes subscribed to a topic, each node includes any other node in its fanout with the probability of `p`, hence, the expected fanout of 
```

File: network/topology/README.md [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+<p align="center">
+  <img src="topicBasedTopology.svg" alt="drawing" width="600"/>
+</p>
+
+### [RandomizedTopology](../../network/topology/randomizedTopology.go)
+
+The randomized-topology is similar to the topic-based topology, except that upon constructing a graph component for a topic, instead of choosing a 
+fixed-size fanout per node, it selects the fanout of the node probabilistically using an _edge probability_, `p`. 
+For the set of nodes subscribed to a topic, each node has any other node in its fanout with the probability of `p`, hence, the expected fanout of 
+a node on the graph component of a topic is `p * x` where `x` is the number of nodes subscribed to the topic. Choosing a proper edge probability
```
Comment (vishalchangrani) [inline network/topology/README.md:53]:
```suggestion
a node on the graph component of a topic is `p * x` where `x` is the number of nodes subscribed to the topic. Choosing an appropriate edge probability 
```

File: network/topology/README.md [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+every node has a fanout size of 3.
+
+<p align="center">
+  <img src="topology.svg" alt="drawing" width="500"/>
+</p>
+
+### [Topology Interface](../../network/topology.go)
+
+At its interface, as a distributed algorithm, the topology receives the (whitelist) of nodes in the system and generates and returns the fanout of the
+node. The current implementations of this topology interface are topic-based topology and randomized topology.
```
Comment (vishalchangrani) [inline network/topology/README.md:30]:
```suggestion
node. The current implementations of this topology interface are topic-based topology and randomized topology. Any future topology implementation must also implement this interface to be able to be plugged into the node.
```

**Review by vishalchangrani — APPROVED** [review]
added some comments but otherwise lgtm

File: network/topology/randomizedTopology.go [hunk 8]
```diff
@@ -15,15 +17,19 @@ import (
 // By random topology we mean a node is connected to any other co-channel nodes with some
 // edge probability.
 type RandomizedTopology struct {
-	me      flow.Identifier             // used to keep identifier of the node
-	state   protocol.State              // used to keep a read only protocol state
-	subMngr network.SubscriptionManager // used to keep track topics the node subscribed to
-	chance  uint64                      // used to translate connectedness probability into a number in [0, 100]
-	rng     random.Rand                 // used as a stateful random number generator to sample edges
+	identifier flow.Identifier             // used to keep identifier of the node
```
Comment (jordanschalm) [inline network/topology/randomizedTopology.go:20]: I think the existing name `me` (or `myID`/`myNodeID` which we use elsewhere) is a more descriptive name. It's more important to know what the identifier refers to than that it is an identifier (we already know that from the type!)

Comment (jordanschalm) [reply network/topology/randomizedTopology.go:20]: Likewise for the other fields where this was changed

**Review by jordanschalm — COMMENTED** [review]
Nice work on the readme and diagrams!

I have one comment on a naming change, otherwise LGTM

File: network/topology/randomizedTopology.go (hunk 8 above)
Comment (yhassanzadeh13) [reply network/topology/randomizedTopology.go]: applied

**Review by jordanschalm — APPROVED** [review]
🎸

**Pull request merged**